kasa/tests/test_dimmer.py

```diff
@@ -0,0 +1,60 @@
+import pytest
+
+from kasa import SmartDimmer
+
+from .conftest import dimmer, handle_turn_on
+
+
+@dimmer
+async def test_set_brightness(dev):
+    await handle_turn_on(dev, False)
+    await dev.set_brightness(99)
+
+    assert dev.brightness == 99
+    assert not dev.is_on
+
+
+@dimmer
+async def test_set_brightness_transition_to_on(dev, mocker):
+    await handle_turn_on(dev, False)
+    query_helper = mocker.spy(SmartDimmer, "_query_helper")
+
+    await dev.set_brightness(99, transition=1000)
+
+    assert dev.brightness == 99
+    assert dev.is_on
+    query_helper.assert_called_with(
+        mocker.ANY,
+        "smartlife.iot.dimmer",
+        "set_dimmer_transition",
+        {"brightness": 99, "duration": 1000},
+    )
+
+
+@dimmer
+async def test_set_brightness_transition_to_off(dev, mocker):
+    await handle_turn_on(dev, True)
+    query_helper = mocker.spy(SmartDimmer, "_query_helper")
+    original_brightness = dev.brightness
+
+    await dev.set_brightness(0, transition=1000)
+
+    assert dev.brightness == original_brightness
```

Comment:
This cannot be correct, right? I think it simply requires `await dev.update()` to update the cache, otherwise it's using the old (pre set_brightness) state.

Reply:
This is correct, when setting to brightness of 0 via a transition it resets the brightness to the original brightness after turning off.

Reply:
I'm not sure how people use this library, so I'm not sure the right API abstraction. Maybe a more logical thing to do would be to disallow a brightness of `0` in `set_brightness()` and to instead add a transition to `turn_off()`.

Reply:
Ah, I see. If you don't mind, add a short note about the behavior.

About the `turn_off` vs `set_brightness(0)`, in my mind it is better to be explicit when the device is turned off. For the bulbs, 0 is also a valid brightness level while still being on, so in my opinion the dimmer should behave the same (setting 0 should set it to the lowest brightness level), `turn_off` should be used to turn the device off.

Reply:
I just tried calling `set_brightness(0)` and it doesn't actually work. Looks like PR #47 was a misguided (or the dimmer switch behaves differently from the bulbs)

```
DEBUG:kasa.protocol:> (63) {"smartlife.iot.dimmer": {"set_brightness": {"brightness": 0}}}
DEBUG:kasa.protocol:< (88) {'smartlife.iot.dimmer': {'set_brightness': {'err_code': -3,
                                             'err_msg': 'invalid argument'}}}
```

Reply:
I pushed another commit that has a different take on how to approach this: https://github.com/python-kasa/python-kasa/commit/9f336ee005f747c36291e34be6d99665041bfced

- Add `transition` support to turn_on and turn_off 
- Disallow brightness 0 in `set_brightness()`, since it doesn't work if there's not a transition and is confusing if there is.

Reply:
Having the `transition` for `turn_{on,off}` sounds good (you need to add ignore for the type error to make mypy pass, see my bulb transition PR).

I just tested my KL60 and for that 0 is the lowest brightness level it has, so I'm assuming the rest of the bulbs do the same. Therefore I would suggest simply coercing the 0 brightness to 1 for dimmers, then the interface would function the same for both bulbs and dimmers (for homeassistant, these are both handled as lights, so having a common interface is preferable).